src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs

```diff
@@ -0,0 +1,2164 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Globalization;
+using System.IO;
+using System.IO.Compression;
+using System.Net;
+using System.Net.Http;
+using System.Net.Security;
+using System.Runtime.InteropServices;
+using System.Security.Authentication;
+using System.Security.Cryptography.X509Certificates;
+using System.Security.Principal;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+using SafeInternetHandle = Interop.WinHttp.SafeInternetHandle;
+
+namespace System.Net.Http
+{
+#if HTTP_DLL
+    internal enum WindowsProxyUsePolicy
+#else
+    public enum WindowsProxyUsePolicy
+#endif
+    {
+        DoNotUseProxy = 0, // Don't use a proxy at all.
+        UseWinHttpProxy = 1, // Use configuration as specified by "netsh winhttp" machine config command. Automatic detect not supported.
+        UseWinInetProxy = 2, // WPAD protocol and PAC files supported.
+        UseCustomProxy = 3 // Use the custom proxy specified in the Proxy property.
+    }
+
+#if HTTP_DLL
+    internal enum AutomaticRedirectionPolicy
+#else
+    public enum AutomaticRedirectionPolicy
+#endif
+    {
+        Never = 0,
+        DisallowHttpsToHttp = 1,
+        Always = 2
+    }
+
+#if HTTP_DLL
+    internal enum CookieUsePolicy
+#else
+    public enum CookieUsePolicy
+#endif
+    {
+        IgnoreCookies = 0,
+        UseInternalCookieStoreOnly = 1,
+        UseSpecifiedCookieContainer = 2
+    }
+
+#if HTTP_DLL
+    internal class WinHttpHandler : HttpMessageHandler
+#else
+    public class WinHttpHandler : HttpMessageHandler
+#endif
+    {
+        private static Interop.WinHttp.WINHTTP_STATUS_CALLBACK s_staticCallback = 
+            new Interop.WinHttp.WINHTTP_STATUS_CALLBACK(WinHttpStatusCallback);
+
+        // TODO:  This looks messy but it is fast. Research a cleaner way
```

Comment:
@stephentoub This is a TODO for a future reminder that we maybe make it cleaner If possible. But this commit will be as-is.